packages/hub/src/lib/parse-safetensors.ts

```diff
@@ -0,0 +1,161 @@
+import type { Credentials, RepoDesignation } from "../types/public";
+import { Counter } from "../utils/Counter";
+import { checkCredentials } from "../utils/checkCredentials";
+import { omit } from "../utils/pick";
+import { toRepoId } from "../utils/toRepoId";
+import { typedEntries } from "../utils/typedEntries";
+import { downloadFile } from "./download-file";
+import { fileExists } from "./file-download-info";
+
+const SINGLE_FILE = "model.safetensors";
+const INDEX_FILE = "model.safetensors.index.json";
+
+type FileName = string;
+
+type TensorName = string;
+type Dtype = "F64" | "F32" | "F16" | "BF16" | "I64" | "I32" | "I16" | "I8" | "U8" | "BOOL";
+
+interface TensorInfo {
+	dtype: Dtype;
+	shape: number[];
+	data_offsets: [number, number];
+}
+
+type FileHeader = Record<TensorName, TensorInfo> & {
+	__metadata__: Record<string, string>;
+};
+
+interface IndexJson {
+	dtype?: string;
+	/// ^there's sometimes a dtype but it looks inconsistent.
+	metadata?: Record<string, string>;
+	/// ^ why the naming inconsistency?
+	weight_map: Record<TensorName, FileName>;
+}
```

Comment:
Let's use `camelCase` for what we expose to the users (here `weight_map` and `data_offsets`), for raw API types there is the `types/api` folder

Also let's prefix types (incl. `IndexJson` and `ShardedHeaders`) with `Tensor`

Reply:
> Also let's prefix types (incl. `IndexJson` and `ShardedHeaders`) with `Tensor`

i'll prefix with `Safetensors` then? (`Tensor` is less relevant at the top-level, it's more like a collection of tensors)

or should i nest everything into a namespace?

Reply:
`Safetensors` yes :+1:

Reply:
feels a bit verbose tbh: ccf2dbea19c7f8726982455819ab89fc5cf0f8c6

Reply:
re. snake_case, i'd rather not transform the data that's actually inside the safetensors files/format, if that's ok